File: Lab9/House.py
import os
from Garage import Garage
from Dimensions import Dimensions
from Communications import Communications

class House:
    def __init__(self, area=None, floors=None, rooms=None, gas=None, water=None, electricity=None, isOpened=None, carsInTheGarage=None, carCapacity=None):
        self.dimensions = Dimensions(area, floors, rooms)
        self.communications = Communications(gas, water, electricity)
        self.garage = Garage(isOpened, carsInTheGarage, carCapacity)
        self.file = open("File.txt", "w")

    def get_dimensions_data(self):
        self.file.write("House Area: " + str(self.dimensions.get_area()) + " sq.m\n")
        self.file.write("Number of Floors: " + str(self.dimensions.get_floors()) + "\n")
        self.file.write("Number of Rooms: " + str(self.dimensions.get_rooms()) + "\n\n")
        self.file.flush()

    def calculate_total_area(self):
        total_area = self.dimensions.calculate_total_area()
        self.file.write("Total area of all rooms in the house: " + str(total_area) + " sq.m\n\n")
        self.file.flush()

    def add_rooms(self):
        if self.dimensions.get_rooms() / self.dimensions.get_floors() > 5:
            self.file.write("Maximum number of rooms (" + str(self.dimensions.get_rooms()) +
                             ") reached for the given number of floors (" + str(self.dimensions.get_floors()) + ")\n")
        else:
            self.dimensions.add_rooms()
            self.file.write("Room added. New number of rooms: " + str(self.dimensions.get_rooms()) + "\n")
        self.file.write("\n")
        self.file.flush()

    def sub_rooms(self):
        if self.dimensions.get_rooms() / self.dimensions.get_floors() <= 1:
            self.file.write("Minimum number of rooms (" + str(self.dimensions.get_rooms()) +
                             ") reached for the given number of floors (" + str(self.dimensions.get_floors()) + ")\n")
        else:
            self.dimensions.sub_rooms()
            self.file.write("Room subtracted. New number of rooms: " + str(self.dimensions.get_rooms()) + "\n")
        self.file.write("\n")
        self.file.flush()

    def turn_on_gas(self):
        self.communications.turn_on_gas()
        self.file.write("Gas turned on.\n\n")
        self.file.flush()

    def turn_off_gas(self):
        self.communications.turn_off_gas()
        self.file.write("Gas turned off.\n\n")
        self.file.flush()

    def turn_on_cold_water(self):
        self.communications.turn_on_cold_water()
        self.file.write("Cold water turned on.\n\n")
        self.file.flush()

    def turn_on_hot_water(self):
        self.communications.turn_on_hot_water()
        self.file.write("Hot water turned on.\n\n")
        self.file.flush()

    def turn_off_water(self):
        self.communications.turn_off_water()
        self.file.write("Water turned off.\n\n")
        self.file.flush()

    def turn_on_electricity(self):
        self.communications.turn_on_electricity()
        self.file.write("Electricity turned on.\n\n")
        self.file.flush()

    def turn_off_electricity(self):
        self.communications.turn_off_electricity()
        self.file.write("Electricity turned off.\n\n")
        self.file.flush()

    def turn_on_all(self):
        self.communications.turn_on_all()
        self.file.write("Gas turned on - " + str(self.communications.get_gas()) + "\n")
        self.file.write("Water turned on - " + str(self.communications.get_water()) + "\n")
        self.file.write("Electricity turned on - " + str(self.communications.get_electricity()) + "\n\n")
        self.file.flush()

    def turn_off_all(self):
        self.communications.turn_off_all()
        self.file.write("Gas turned off - " + str(self.communications.get_gas()) + "\n")
        self.file.write("Water turned off - " + str(self.communications.get_water()) + "\n")
        self.file.write("Electricity turned off - " + str(self.communications.get_electricity()) + "\n\n")
        self.file.flush()

    def get_communications_data(self):
        self.file.write("Gas state - " + str(self.communications.get_gas()) + "\n")
        self.file.write("Water state - " + str(self.communications.get_water()) + "\n")
        self.file.write("Electricity state - " + str(self.communications.get_electricity()) + "\n\n")
        self.file.flush()

    def open_garage_remote(self):
        self.garage.open_garage_remote()
        self.file.write("Garage opened remotely: " + str(self.garage.is_opened()) + "\n\n")
        self.file.flush()

    def open_garage_by_key(self):
        self.garage.open_garage_by_key()
        self.file.write("Garage opened by key: " + str(self.garage.is_opened()) + "\n\n")
        self.file.flush()

    def close_garage(self):
        self.garage.close_garage()
        self.file.write("Garage closed: " + str(self.garage.is_opened()) + "\n\n")
        self.file.flush()

    def get_garage_data(self):
        self.file.write("Garage state: " + str(self.garage.is_opened()) + "\n")
        self.file.write("Number of parked cars: " + str(self.garage.get_cars_in_the_garage()) + "\n")
        self.file.write("Maximum car capacity: " + str(self.garage.get_car_capacity()) + "\n\n")
        self.file.flush()

    def park_car(self):
        if not self.garage.is_opened():
            self.file.write("Garage is closed. Unable to park a car.\n")
        elif self.garage.get_cars_in_the_garage() + 1 > self.garage.get_car_capacity():
            self.file.write("No space in the garage. Parked " + str(self.garage.get_cars_in_the_garage()) +
                            " cars out of possible " + str(self.garage.get_car_capacity()) + "\n")
        else:
            self.garage.park_car()
            self.file.write("Car parked. New number of parked cars " +
                            str(self.garage.get_cars_in_the_garage()) + " cars out of possible " +
                            str(self.garage.get_car_capacity()) + "\n")
        self.file.write("\n")
        self.file.flush()

    def take_car(self):
        if not self.garage.is_opened():
            self.file.write("Garage is closed. Unable to take a car.\n")
        elif self.garage.get_cars_in_the_garage() <= 0:
            self.file.write("No cars in the garage.\n")
        else:
            self.garage.take_car()
            self.file.write("Car taken from the garage. New number of cars: " +
                            str(self.garage.get_cars_in_the_garage()) + " cars out of possible " +
                            str(self.garage.get_car_capacity()) + "\n")
        self.file.write("\n")
        self.file.flush()

    def add_car_capacity(self):
        self.garage.add_car_capacity()
        self.file.write("New garage car capacity: " +
                        str(self.garage.get_car_capacity()) + "\n\n")
        self.file.flush()

    def get_dimensions(self):
        return self.dimensions

    def close_file(self):
        self.file.close()

File: Lab2/src/ki304/kostiuk/Lab2/House.java
package ki304.kostiuk.Lab2;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.PrintWriter;

import ki304.kostiuk.Lab2.Communications.WaterType;
import ki304.kostiuk.Lab2.Garage.GarageType;

public class House {
	private Dimensions dimensions;
    private Communications communications;
    private Garage garage;
    private PrintWriter file;
    
    /**
     * Конструктор за замовчуванням. Створює об'єкт будинку і ініціалізує об'єкти розмірів, комунікацій та гаражу.
     * Виконує ініціалізацію файлу для журналу.
     * @throws FileNotFoundException якщо не вдається створити файл для журналу.
     */
    public House() throws FileNotFoundException {
        dimensions = new Dimensions();
        communications = new Communications();
        garage = new Garage();
        file = new PrintWriter(new File("File.txt"));
    }


    /**
     * Конструктор, який дозволяє вказати розміри будинку, стан газу, води, електрики та гаражу.
     * Виконує ініціалізацію об'єктів розмірів, комунікацій та гаражу. Також ініціалізує файл для журналу.
     * @param area площа будинку.
     * @param floors кількість поверхів в будинку.
     * @param rooms кількість кімнат в будинку.
     * @param gas стан газу в будинку (true - увімкнено, false - вимкнено).
     * @param water стан води в будинку (HOT або COLD).
     * @param electricity стан електрики в будинку (true - увімкнено, false - вимкнено).
     * @param isOpened стан гаражу (OPEN або CLOSED).
     * @param carsInTheGarage кількість припаркованих автомобілів в гаражі.
     * @param carCapacity максимальна кількість автомобілів, що можуть бути припарковані в гаражі.
     * @throws FileNotFoundException якщо не вдається створити файл для журналу.
     */
    public House(double area, int floors, int rooms, boolean gas, WaterType water, boolean electricity, 
    		GarageType isOpened, int carsInTheGarage, int carCapacity ) throws FileNotFoundException {
    	dimensions = new Dimensions(area, floors, rooms);
    	communications = new Communications(gas, water, electricity );
    	garage = new Garage (isOpened, carsInTheGarage, carCapacity  );
    	file = new PrintWriter(new File("File.txt"));
    }
    
    /**
     * Конструктор, який дозволяє вказати розміри будинку.
     * Виконує ініціалізацію об'єкта розмірів, а також ініціалізацію файлу для журналу.
     * @param area площа будинку.
     * @param floors кількість поверхів в будинку.
     * @param rooms кількість кімнат в будинку.
     * @throws FileNotFoundException якщо не вдається створити файл для журналу.
     */
    public House(double area, int floors, int rooms) throws FileNotFoundException {
    	dimensions = new Dimensions(area, floors, rooms);
    	communications = new Communications();
    	garage = new Garage();
    	file = new PrintWriter(new File("File.txt"));
    }
    
    /**
     * Конструктор, який дозволяє вказати стан газу, води та електрики в будинку.
     * Виконує ініціалізацію об'єкта комунікацій та ініціалізацію файлу для журналу.
     * @param gas стан газу в будинку (true - увімкнено, false - вимкнено).
     * @param water стан води в будинку (HOT або COLD).
     * @param electricity стан електрики в будинку (true - увімкнено, false - вимкнено).
     * @throws FileNotFoundException якщо не вдається створити файл для журналу.
     */
    public House(boolean gas, WaterType water, boolean electricity) throws FileNotFoundException {
    	dimensions = new Dimensions();
    	communications = new Communications(gas, water, electricity );
    	garage = new Garage();
    	file = new PrintWriter(new File("File.txt"));
    }
    
    /**
     * Конструктор, який дозволяє вказати стан гаражу та максимальну кількість автомобілів, які можуть бути припарковані.
     * Виконує ініціалізацію об'єкта гаражу та ініціалізацію файлу для журналу.
     * @param isOpened стан гаражу (OPEN або CLOSED).
     * @param carsInTheGarage кількість припаркованих автомобілів в гаражу.
     * @param carCapacity максимальна кількість автомобілів, що можуть бути припарковані в гаражі.
     * @throws FileNotFoundException якщо не вдається створити файл для журналу.
     */
    public House(GarageType isOpened, int carsInTheGarage, int carCapacity) throws FileNotFoundException {
    	dimensions = new Dimensions();
    	garage = new Garage (isOpened, carsInTheGarage, carCapacity  );
    	communications = new Communications();
    	file = new PrintWriter(new File("File.txt"));
    }
    
/**
 * Виводить інформацію про розміри будинку (площу, кількість поверхів та кімнат)
 * у вигляді текстових рядків у вказаний вихідний потік.
 */
public void getDimensionsData() {
	file.println("Площа будинку: " + dimensions.getArea() + " кв. м");
	file.println("Кількість поверхів: " + dimensions.getFloors());
	file.println("Кількість кімнат: " + dimensions.getRooms());
	file.println();
	file.flush();
}

/**
 * Обчислює та виводить загальну площу усіх кімнат у будинку у вихідний потік.
 * Результат обчислення виводиться як рядок у вигляді "Загальна площа усіх кімнат у будинку : [результат] кв. м".
 */
public void calculateTotalArea() {
	dimensions.calculateTotalArea();
	file.println("Загальна площа усіх кімнат у будинку : " + dimensions.calculateTotalArea());
	file.println();
	file.flush();
}

/**
 * Метод для додавання кімнат у будинку. Перевіряє, чи не досягнуто максимальної
 * кількості кімнат для даної кількості поверхів, і виводить відповідне повідомлення у файл.
 * Якщо максимальна кількість кімнат не досягнута, то метод додає одну кімнату та
 * виводить нову кількість кімнат у файл.
 */
public void addRooms(){
	if(dimensions.getRooms()/dimensions.getFloors() > 5 ) {
		file.println("Досягнуто максимальної к-сті кімнат (" + dimensions.getRooms() + ") для даної к-сті поверхів (" + 
				dimensions.getFloors()+")" );
	}
	else {
		dimensions.addRooms();
		file.println("Додано кімнату. Нова к-сть кімнат : " + dimensions.getRooms());
	}
	file.println();
	file.flush();
}

/**
 * Метод для зменшення кількості кімнат у будівлі. Перевіряє, чи не досягнуто мінімальної
 * кількості кімнат для даної кількості поверхів, і виводить відповідне повідомлення у файл.
 * Якщо мінімальна кількість кімнат не досягнута, то метод зменшує кількість кімнат на одну.
 * Після виконання операції, записує результат у файл.
 */
public void subRooms() {
	if(dimensions.getRooms() / dimensions.getFloors() <= 1) {
		file.println("Досягнуто мінімальної к-сті кімнат (" + dimensions.getRooms() + ") для даної к-сті поверхів (" + 
				dimensions.getFloors()+")" );
	}
	else {
		dimensions.subRooms();
		file.println("Віднято кімнату. Нова к-сть кімнат : " + dimensions.getRooms());
	}
	file.println();
	file.flush();
}

/**
 * Увімкнути газ у будинку.
 * Цей метод включає газ у комунікаціях будинку.
 */
public void turnOnGas() {
	communications.turnOnGas();
	file.println("Газ увімкнено.");
	file.println();
	file.flush();
}

/**
 * Вимкнути газ у будинку.
 * Цей метод вимикає газ у комунікаціях будинку .
 */
public void turnOffGas() {
	communications.turnOffGas();;
	file.println("Газ вимкнено.");
	file.println();
	file.flush();
}

/**
 * Увімкнути холодну воду у будинку.
 * Цей метод включає холодну воду у комунікаціях будинку.
 */
public void turnOnColdWater() {
	communications.turnOnColdWater();
	file.println("Холодну воду увімкнено.");
	file.println();
	file.flush();
}

/**
 * Увімкнути теплу воду у будинку.
 * Цей метод включає теплу воду у комунікаціях будинку.
 */
public void turnOnHotWater() {
	communications.turnOnHotWater();
	file.println("Теплу воду увімкнено.");
	file.println();
	file.flush();
}

/**
 * Вимкнути воду у будинку.
 * Цей метод виключає воду у комунікаціях будинку.
 */
public void turnOffWater() {
	communications.turnOffWater();
	file.println("Теплу воду увімкнено.");
	file.println();
	file.flush();
}

/**
 * Увімкнути електрику у будинку.
 * Цей метод включає електрику у комунікаціях будинку.
 */
public void turnOnElectricity() {
	communications.turnOnElectricity();
	file.println("Електрику увімкнено.");
	file.println();
	file.flush();
}

/**
 * Вимкнути електрику у будинку.
 * Цей метод виключає електрику у комунікаціях будинку.
 */
public void turnOffElectricity() {
	communications.turnOffElectricity();
	file.println("Електрику вимкнено.");
	file.println();
	file.flush();
}

/**
 * Увімкнути всі комунікації в будинку (газ, воду, електрику).
 */
public void turnOnAll() {
    communications.turnOnAll();
    file.println("Увімкнено газ - " + communications.getGas());
    file.println("Увімкнено воду - " + communications.getWater());
    file.println("Увімкнено електрику - " + communications.getElectricity());
    file.println();
    file.flush();
}

/**
 * Вимкнути всі комунікації в будинку (газ, воду, електрику).
 */
public void turnOffAll() {
    communications.turnOffAll();
    file.println("Вимкнено газ - " + communications.getGas());
    file.println("Вимкнено воду - " + communications.getWater());
    file.println("Вимкнено електрику - " + communications.getElectricity());
    file.println();
    file.flush();
}

/**
 * Отримати інформацію про стан всіх комунікацій в будинку (газ, вода, електрика).
 */
public void getCommunicationsData() {
    file.println("Стан газу - " + communications.getGas());
    file.println("Стан води - " + communications.getWater());
    file.println("Стан електрики - " + communications.getElectricity());
    file.println();
    file.flush();
}

/**
 * Відкрити гараж дистанційно.
 */
public void openGarageRemote() {
    garage.openGarageRemote();
    file.println("Гараж відкрито дистанційно : " + garage.getIsOpened());
    file.println();
    file.flush();
}

/**
 * Відкрити гараж за допомогою ключа.
 */
public void openGarageByKey() {
    garage.openGarageByKey();
    file.println("Гараж відкрито ключем : " + garage.getIsOpened());
    file.println();
    file.flush();
}

/**
 * Закрити гараж.
 */
public void closeGarage() {
    garage.closeGarage();
    file.println("Гараж закрито : " + garage.getIsOpened());
    file.println();
    file.flush();
}

/**
 * Отримати інформацію про стан гаражу та кількість автомобілів, що знаходяться в гаражі.
 */
public void getGarageData() {
    file.println("Стан гаражу : " + garage.getIsOpened());
    file.println("Кількість припаркованих авто : " + garage.getCarsInTheGarage());
    file.println("Максимальна кількість вмістимих авто : " + garage.getCarCapacity());
    file.println();
    file.flush();
}

/**
 * Припаркувати автомобіль у гаражу.
 */
public void parkCar() {
    if (!garage.isGarageOpened()) {
        file.println("Гараж закритий. Неможливо припаркувати авто.");
    } else if (garage.getCarsInTheGarage() + 1 > garage.getCarCapacity()) {
        file.println("У гаражі немає місця. Припарковано " + garage.getCarsInTheGarage() + " авто з можливих " +
                garage.getCarCapacity());
    } else {
        garage.parkCar();
        file.println("Авто припарковано. Нова кількість припаркованих авто " + garage.getCarsInTheGarage() +
                " авто з можливих " + garage.getCarCapacity());
    }
    file.println();
    file.flush();
}

/**
 * Взяти автомобіль з гаражу.
 */
public void takeCar() {
    if (!garage.isGarageOpened()) {
        file.println("Гараж закритий. Неможливо взяти авто.");
    } else if (garage.getCarsInTheGarage() <= 0) {
        file.println("У гаражі немає машин.");
    } else {
        garage.takeCar();
        file.println("Авто взято з гаражу. Нова кількість автомобілів: " + garage.getCarsInTheGarage() +
        " авто з можливих " + garage.getCarCapacity());
    }
    file.println();
    file.flush();
}

/**
 * Збільшити максимальну кількість автомобілів, які можуть бути припарковані у гаражі.
 */
public void addCarCapacity() {
    garage.addCarCapacity();
    file.println("Нова кількість авто вмістимості гаражу : " + garage.getCarCapacity());
    file.println();
    file.flush();
}

}
